### PR TITLE
Fix #261:Add support for traceparent headers in DefaultRestClient

### LINKED ISSUE #261: Add support for traceparent headers in DefaultRestClient

### PR DESCRIPTION
**Pull request by jandusil**

### CODE REVIEW

File: rest-client-base/src/main/java/com/wultra/core/rest/client/base/DefaultRestClient.java [hunk 1]
```diff
@@ -217,6 +220,32 @@ private void initializeWebClient() throws RestClientException {
...
+     * @param headers The HttpHeaders object where the traceparent header will be added.
+     *                This object is expected to be non-null.
+     */
+    private static void injectTraceparentHeader(final HttpHeaders headers) {
+        final Optional<SpanContext> spanContext = Optional.ofNullable(Span.current().getSpanContext());
+
+        if (spanContext.isPresent() && headers != null) {
+            final String traceId = spanContext.map(SpanContext::getTraceId).orElse(null);
+            final String spanId = spanContext.map(SpanContext::getSpanId).orElse(null);
+            final TraceFlags traceFlags = spanContext.map(SpanContext::getTraceFlags).orElse(null);
```
Comment (banterCZ) [inline rest-client-base/src/main/java/com/wultra/core/rest/client/base/DefaultRestClient.java:237]: Why so complicated solution with `Optional`?
```suggestion
        if (headers == null) {
            return;
        }

        final SpanContext spanContext = Span.current().getSpanContext();
        
        if (spanContext != null) {
            final String traceId = spanContext.getTraceId();
            final String spanId = spanContext.getSpanId();
            final TraceFlags traceFlags = spanContext.getTraceFlags();Flags = spanContext.map(SpanContext::getTraceFlags).orElse(null);
```

File: rest-client-base/src/main/java/com/wultra/core/rest/client/base/DefaultRestClient.java [hunk 2]
```diff
@@ -217,6 +220,32 @@ private void initializeWebClient() throws RestClientException {
         webClient = builder.baseUrl(config.getBaseUrl()).clientConnector(connector).build();
     }
 
+    /**
+     * Injects a traceparent header into the provided HttpHeaders object.
+     * The header is composed of the trace ID and span ID obtained from the current SpanContext.
+     * The header is only added if both trace ID and span ID are present.
+     *
+     * @param headers The HttpHeaders object where the traceparent header will be added.
+     *                This object is expected to be non-null.
```
Comment (banterCZ) [inline rest-client-base/src/main/java/com/wultra/core/rest/client/base/DefaultRestClient.java:229]: May we be more strict and add assertation/valiation?

File: rest-client-base/src/main/java/com/wultra/core/rest/client/base/RestClientConfiguration.java [hunk 3]
```diff
@@ -41,6 +41,7 @@ public class RestClientConfiguration {
     private String baseUrl;
     private MediaType contentType = MediaType.APPLICATION_JSON;
     private MediaType acceptType = MediaType.APPLICATION_JSON;
+    protected static final String TRACEPARENT_HEADER_KEY = "traceparent";
```
Comment (banterCZ) [inline rest-client-base/src/main/java/com/wultra/core/rest/client/base/RestClientConfiguration.java:44]: Why `protected` and placed in the configuration class?

**Review by banterCZ — COMMENTED** [review]
Not covered by tests.

**Comment by jandusil** [comment]
Based on this [blog](https://azizulhaq-ananto.medium.com/how-to-handle-logs-and-tracing-in-spring-webflux-and-microservices-a0b45adc4610) I aimed to implement a way to propagate the traceId through `Reactor` context. 
However, the blog was based on Sleuth solution, and it may be that the opentelemetry handles the non-blocking environment well without issues. Unfortunately, I have not found any evidence that would support nor contradict this claim.

Can you please take a look at this solution? If it is accepted, I will add some tests as well.

File: rest-client-base/src/main/java/com/wultra/core/rest/client/base/TraceparentFilterFunction.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    public Mono<ClientResponse> filter(final ClientRequest request, final ExchangeFunction next) {
+        return next.exchange(addTraceparentHeader(request)).contextWrite(
+                context ->
+                {
+                    final Span currentSpan = Span.current();
+                    if (currentSpan != null) {
+                        final SpanContext spanContext = currentSpan.getSpanContext();
+                        if (spanContext != null) {
+                            final Context contextTmp = context.put("traceId", spanContext.getTraceId());
+                            MDC.put("traceId", spanContext.getTraceId());
```
Comment (romanstrobl) [inline rest-client-base/src/main/java/com/wultra/core/rest/client/base/TraceparentFilterFunction.java:56]: We put the `traceId` into the MDC, however we do not put the `spanId` into MDC. Can you explain what is this used for and why only `traceId` is added?

**Review by romanstrobl — APPROVED** [review]
The feature seems to be implemented based on the W3C specification.

I would suggest to make this feature optional (e.g. allow disabling adding the header using a property).

File: rest-client-base/src/main/java/com/wultra/core/rest/client/base/TraceparentFilterFunction.java (hunk 4 above)
Comment (jandusil) [reply rest-client-base/src/main/java/com/wultra/core/rest/client/base/TraceparentFilterFunction.java:56]: The way I understand it, the `spanId` will be changed as soon as the server picks up the client call anyway as it is another subprocess in the workflow. The main addition is to be able to propagate `traceId` so the whole workflow gets grouped by it in the logs.

**Comment by jandusil** [comment]
More info about the changes in the issue https://github.com/wultra/lime-java-core/issues/261

**Review by zcgandcomp — APPROVED** [review]
Looks OK.

**Pull request merged**